core/src/main/java/io/grpc/Metadata.java

```diff
@@ -694,4 +681,35 @@ public String toString() {
       }
     }
   }
+
+  private static class ValueIterable<T> implements Iterable<T> {
+    private final Key<T> key;
+    private final Iterable<MetadataEntry> entries;
+
+    public ValueIterable(Key<T> key, Iterable<MetadataEntry> entries) {
+      this.key = key;
+      this.entries = entries;
+    }
+
+    @Override
+    public Iterator<T> iterator() {
+      final Iterator<MetadataEntry> iterator = entries.iterator();
+      class ValueIterator implements Iterator<T> {
+        @Override public boolean hasNext() {
+          return iterator.hasNext();
+        }
+
+        @Override public T next() {
```

Comment:
I normally do that, but in this case we get it for "free" because we are backed by another iterator. I expect that other iterator to throw, per the Iterator interface.